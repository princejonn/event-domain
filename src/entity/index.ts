export * from "./Aggregate";
export * from "./Saga";
export * from "./View";
