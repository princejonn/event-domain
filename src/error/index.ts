export * from "./AggregateAlreadyCreatedError";
export * from "./AggregateDestroyedError";
export * from "./AggregateNotCreatedError";
export * from "./AggregateNotDestroyedError";
export * from "./CausationMissingEventsError";
export * from "./CommandSchemaValidationError";
export * from "./ConcurrencyError";
export * from "./DomainError";
export * from "./EventStreamRevisionError";
export * from "./HandlerNotRegisteredError";
export * from "./IllegalEntityChangeError";
export * from "./MessageTypeError";
export * from "./MongoDuplicateKeyError";
export * from "./MongoNotUpdatedError";
export * from "./SagaAlreadyCreatedError";
export * from "./SagaCausationError";
export * from "./SagaDestroyedError";
export * from "./SagaNotCreatedError";
export * from "./SagaNotInStoreError";
export * from "./SagaRevisionError";
export * from "./ViewAlreadyCreatedError";
export * from "./ViewCausationError";
export * from "./ViewDestroyedError";
export * from "./ViewNotCreatedError";
export * from "./ViewRevisionError";
