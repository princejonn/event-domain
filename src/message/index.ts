export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
