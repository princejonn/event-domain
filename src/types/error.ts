import { LindormErrorOptions } from "@lindorm-io/errors";

export interface DomainErrorOptions extends LindormErrorOptions {
  permanent: boolean;
}
