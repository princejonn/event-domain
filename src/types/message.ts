import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
