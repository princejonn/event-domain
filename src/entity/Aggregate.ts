import { AGGREGATE_MERGE_STATE_SCHEMA, getApplyEventSchema } from "../schema";
import { AggregateEventHandler } from "../handler";
import { Command, DomainEvent } from "../message";
import { Logger } from "@lindorm-io/winston";
import { assertCamelCase, validateSchemaAsync, validateSchemaSync } from "../util";
import { cloneDeep, find, merge, set } from "lodash";
import {
  AggregateData,
  AggregateEventHandlerContext,
  AggregateIdentifier,
  AggregateOptions,
  AnyType,
  GenericRecord,
  IAggregate,
} from "../types";
import {
  AggregateDestroyedError,
  AggregateNotDestroyedError,
  HandlerNotRegisteredError,
  IllegalEntityChangeError,
  MessageTypeError,
} from "../error";

export class Aggregate<State> implements IAggregate<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _eventHandlers: Array<AggregateEventHandler<State>>;
  private readonly _events: Array<DomainEvent>;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;
  private _destroying: boolean;
  private _numberOfLoadedEvents: number;

  private readonly logger: Logger;

  constructor(options: AggregateOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["Aggregate"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._destroyed = false;
    this._destroying = false;
    this._eventHandlers = options.eventHandlers || [];
    this._events = [];
    this._numberOfLoadedEvents = 0;
    this._state = {};
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get events(): Array<DomainEvent> {
    return this._events;
  }
  public set events(_) {
    throw new IllegalEntityChangeError();
  }

  public get numberOfLoadedEvents(): number {
    return this._numberOfLoadedEvents;
  }
  public set numberOfLoadedEvents(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(_) {
    throw new IllegalEntityChangeError();
  }

  public async apply(
    causation: Command,
    name: string,
    data?: Record<string, any>,
  ): Promise<void> {
    this.logger.debug("applying command", { causation, name, data });

    await this.handleEvent(
      new DomainEvent(
        {
          aggregate: { id: this.id, name: this.name, context: this.context },
          data: data,
          name: name,
        },
        causation,
      ),
    );
  }

  public async load(event: DomainEvent): Promise<void> {
    this.logger.debug("loading event", { event });

    await this.handleEvent(event);
    this._numberOfLoadedEvents += 1;
  }

  private async handleEvent(event: DomainEvent): Promise<void> {
    this.logger.debug("handling event", { event });

    if (!(event instanceof DomainEvent)) {
      throw new MessageTypeError(event, DomainEvent);
    }

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    const destroying = this._destroying;

    await validateSchemaAsync(
      getApplyEventSchema({
        id: this.id,
        name: this.name,
        context: this.context,
      }),
      {
        name: event.name,
        aggregate: event.aggregate,
        data: event.data,
        causationId: event.causationId,
      },
    );

    const eventHandler: AggregateEventHandler<State> = find(
      this._eventHandlers,
      (handler) => handler.eventName === event.name,
    );

    if (!(eventHandler instanceof AggregateEventHandler)) {
      throw new HandlerNotRegisteredError();
    }

    const context: AggregateEventHandlerContext<State> = {
      event: event,
      state: cloneDeep(this._state),
      logger: this.logger.createChildLogger(["AggregateEventHandler"]),
      destroy: this.destroy.bind(this),
      destroyNext: this.destroyNext.bind(this),
      mergeState: this.mergeState.bind(this),
      setState: this.setState.bind(this),
    };

    await eventHandler.handler(context);

    if (destroying && !this._destroyed) {
      throw new AggregateNotDestroyedError();
    }

    this._events.push(event);

    this.logger.debug("successfully handled event", { event });
  }

  private destroy(): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroyed = true;
    this._destroying = false;
  }

  private destroyNext(): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroying = true;
  }

  private mergeState(data: Record<string, any>): void {
    validateSchemaSync(AGGREGATE_MERGE_STATE_SCHEMA, { data });

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    merge(this._state, data);
  }

  private setState(path: string, value: AnyType): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    set(this._state, path, value);
  }

  public toJSON(): AggregateData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      destroyed: this.destroyed,
      events: cloneDeep(this.events),
      numberOfLoadedEvents: cloneDeep(this.numberOfLoadedEvents),
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: AggregateIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
