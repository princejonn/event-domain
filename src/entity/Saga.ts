import { Command, DomainEvent, Timeout } from "../message";
import {
  ISaga,
  SagaIdentifier,
  SagaOptions,
  AnyType,
  GenericRecord,
  SagaData,
} from "../types";
import { IllegalEntityChangeError, MessageTypeError, SagaDestroyedError } from "../error";
import { SAGA_DISPATCH_SCHEMA, SAGA_UPDATE_STATE_SCHEMA } from "../schema";
import { assertCamelCase, validateSchemaSync } from "../util";
import { cloneDeep, merge, set } from "lodash";
import { Logger } from "@lindorm-io/winston";

export class Saga<State> implements ISaga<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _causationList: Array<string>;
  private readonly _messagesToDispatch: Array<Command | DomainEvent>;
  private readonly _revision: number;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;

  private readonly logger: Logger;

  constructor(options: SagaOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["Saga"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._causationList = options.causationList || [];
    this._destroyed = options.destroyed || false;
    this._messagesToDispatch = options.messagesToDispatch || [];
    this._revision = options.revision || 0;
    this._state = options.state || {};
  }

  public get causationList(): Array<string> {
    return this._causationList;
  }
  public set causationList(_) {
    throw new IllegalEntityChangeError();
  }

  public get messagesToDispatch(): Array<Command> {
    return this._messagesToDispatch;
  }
  public set messagesToDispatch(_) {
    throw new IllegalEntityChangeError();
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get revision(): number {
    return this._revision;
  }
  public set revision(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(_) {
    throw new IllegalEntityChangeError();
  }

  public destroy(): void {
    this.logger.debug("destroying");

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._destroyed = true;
  }

  public dispatch(causation: DomainEvent, command: Command): void {
    this.logger.debug("dispatching command", { causation, command });

    validateSchemaSync(SAGA_DISPATCH_SCHEMA, { command, causation });

    if (!(command instanceof Command)) {
      throw new MessageTypeError(command, Command);
    }

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._messagesToDispatch.push(
      new Command(
        merge(
          {
            aggregate: {
              context: this.context,
            },
          },
          command,
        ),
        causation,
      ),
    );
  }

  public mergeState(data: Record<string, any>): void {
    this.logger.debug("merging state", { data });

    validateSchemaSync(SAGA_UPDATE_STATE_SCHEMA, { data });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    merge(this._state, data);
  }

  public setState(path: string, value: AnyType): void {
    this.logger.debug("setting state", { path, value });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    set(this._state, path, value);
  }

  public timeout(
    causation: DomainEvent,
    name: string,
    data: Record<string, any>,
    delay: number,
  ): void {
    this.logger.debug("dispatching timeout", { causation, name, data, delay });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._messagesToDispatch.push(
      new Timeout(
        {
          name,
          data,
          aggregate: {
            id: this.id,
            name: this.name,
            context: this.context,
          },
          delay,
        },
        causation,
      ),
    );
  }

  public toJSON(): SagaData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      causationList: cloneDeep(this.causationList),
      messagesToDispatch: cloneDeep(this.messagesToDispatch),
      destroyed: this.destroyed,
      revision: this.revision,
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: SagaIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
