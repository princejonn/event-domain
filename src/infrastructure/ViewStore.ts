import { DomainEvent } from "../message";
import { Logger } from "@lindorm-io/winston";
import { StoreType } from "../enum";
import { View } from "../entity";
import { ViewStoreInMemory } from "./in-memory";
import { ViewStoreMongo } from "./mongo";
import {
  IViewStore,
  ViewIdentifier,
  ViewStoreCollectionOptions,
  ViewStoreOptions,
  ViewStoreQuery,
} from "../types";

export class ViewStore<State> implements IViewStore<State> {
  private readonly logger: Logger;
  private store: IViewStore<State>;

  constructor(logger: Logger, options?: ViewStoreOptions) {
    this.logger = logger.createChildLogger(["ViewStore"]);

    const type = options?.type || StoreType.MEMORY;

    switch (type) {
      case StoreType.CUSTOM:
        this.store = options.custom;
        break;

      case StoreType.MEMORY:
        this.store = new ViewStoreInMemory(this.logger);
        break;

      case StoreType.MONGO:
        this.store = new ViewStoreMongo(this.logger, options.mongo);
        break;

      default:
        throw new Error("Invalid Store Type");
    }
  }

  public async save(
    view: View<State>,
    causation: DomainEvent,
    options?: ViewStoreCollectionOptions<State>,
  ): Promise<View<State>> {
    try {
      const result = await this.store.save(view, causation, options);

      this.logger.debug("successfully saved view", { view: View.identifier(view) });

      return result;
    } catch (err) {
      this.logger.error("failed to save view", { view: View.identifier(view) });

      throw err;
    }
  }

  public async load(
    viewIdentifier: ViewIdentifier,
    options?: ViewStoreCollectionOptions<State>,
  ): Promise<View<State>> {
    try {
      const result = await this.store.load(viewIdentifier, options);

      if (!result) {
        this.logger.debug("successfully created view", {
          view: View.identifier(viewIdentifier),
        });

        return new View(viewIdentifier, this.logger);
      }

      this.logger.debug("successfully loaded view", {
        view: View.identifier(viewIdentifier),
      });

      return result;
    } catch (err) {
      this.logger.error("failed to load view", { view: View.identifier(viewIdentifier) });

      throw err;
    }
  }

  public async query(query: ViewStoreQuery): Promise<Array<View<State>>> {
    try {
      const result = await this.store.query(query);

      this.logger.debug("successfully queried view", { query, length: result.length });

      return result;
    } catch (err) {
      this.logger.error("failed to query view", { query });

      throw err;
    }
  }
}
