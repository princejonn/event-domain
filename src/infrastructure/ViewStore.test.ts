import MockDate from "mockdate";
import { ViewStore } from "./ViewStore";
import { logger } from "../test";
import { Command, DomainEvent } from "../message";
import { View } from "../entity";
import { ViewCausationError, ViewRevisionError } from "../error";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("ViewStore", () => {
  const aggregateIdentifier = {
    id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
    name: "aggregateName",
    context: "aggregateContext",
  };
  const viewIdentifier = {
    id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
    name: "viewName",
    context: "viewContext",
  };

  let viewStore: ViewStore<any>;
  let event: DomainEvent;

  beforeEach(() => {
    viewStore = new ViewStore(logger);

    event = new DomainEvent(
      {
        id: "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
        name: "eventName",
        aggregate: aggregateIdentifier,
        data: { eventData: true },
      },
      new Command({
        id: "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
        name: "commandName",
        aggregate: aggregateIdentifier,
      }),
    );
  });

  describe("constructor", () => {
    it("should create a view store", () => {
      expect(new ViewStore(logger)).toStrictEqual(expect.any(ViewStore));
    });

    it("should throw if view store type is erroneous", () => {
      expect(
        () =>
          new ViewStore({
            // @ts-ignore
            type: "wrong",
          }),
      ).toThrow();
    });
  });

  describe("save", () => {
    it("should save a view", async () => {
      await expect(
        viewStore.save(
          new View(
            {
              ...viewIdentifier,
              state: { data: true },
            },
            logger,
          ),
          event,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should throw if causation already exists in list", async () => {
      await expect(
        viewStore.save(
          new View(
            {
              ...viewIdentifier,
              causationList: [event.id],
            },
            logger,
          ),
          event,
        ),
      ).rejects.toThrow(ViewCausationError);
    });

    it("should throw if revision is wrong", async () => {
      const view = await viewStore.save(
        new View(
          {
            ...viewIdentifier,
            causationList: [],
          },
          logger,
        ),
        event,
      );

      const event2 = new DomainEvent(
        {
          id: "55b4e3ef-3a9c-4f72-9c6a-765ed59bc1b0",
          name: "eventName2",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        new Command({
          id: "a38165e1-1336-45cf-a428-377231ed9af1",
          name: "commandName2",
          aggregate: aggregateIdentifier,
        }),
      );

      await expect(
        viewStore.save(
          new View(
            {
              id: view.id,
              name: view.name,
              context: view.context,
              causationList: view.causationList,
              destroyed: view.destroyed,
              metaState: view.metaState,
              revision: 99,
              state: view.state,
            },
            logger,
          ),
          event2,
        ),
      ).rejects.toThrow(ViewRevisionError);
    });
  });

  describe("load", () => {
    it("should load an existing view", async () => {
      await viewStore.save(
        new View(
          {
            ...viewIdentifier,
            state: { exists: true },
          },
          logger,
        ),
        event,
      );

      await expect(viewStore.load(viewIdentifier)).resolves.toMatchSnapshot();
    });

    it("should load a new view", async () => {
      await expect(viewStore.load(viewIdentifier)).resolves.toMatchSnapshot();
    });
  });
});
