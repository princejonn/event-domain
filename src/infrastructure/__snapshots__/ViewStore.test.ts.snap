// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewStore load should load a new view 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {},
  "name": "viewName",
  "revision": 0,
  "state": Object {},
}
`;

exports[`ViewStore load should load an existing view 1`] = `
Object {
  "causationList": Array [
    "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
  ],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {},
  "name": "viewName",
  "revision": 1,
  "state": Object {
    "exists": true,
  },
}
`;

exports[`ViewStore save should save a view 1`] = `
Object {
  "causationList": Array [
    "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
  ],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {},
  "name": "viewName",
  "revision": 1,
  "state": Object {
    "data": true,
  },
}
`;
